commit 495be913656c9652d9fd0b0cc79a898c0dde6173
Author: Quentin Perret <quentin.perret@arm.com>
Date:   Thu Feb 14 14:52:15 2019 +0000

    ANDROID: eas: Add CPU capacity parsing from sysfs
    
    In Android 4.9 and before, CPU capacities can be read from the sched
    domain debug interface exposed by EAS. But more recent kernels use the
    mainline alternative using sysfs. In order to support both, attempt to
    read data from sysfs whenever the sched domain debug files aren't
    present.
    
    While at it, ensure to free() the buffer used to read files once not
    needed any more.
    
    Signed-off-by: Quentin Perret <quentin.perret@arm.com>
    Change-Id: I9b52b5f89ebd642e18a30bc7e26447be6aa83e7f

diff --git a/testcases/kernel/sched/eas/util.c b/testcases/kernel/sched/eas/util.c
index 0f0ef13fa..b6c051904 100644
--- a/testcases/kernel/sched/eas/util.c
+++ b/testcases/kernel/sched/eas/util.c
@@ -5,6 +5,7 @@
  */
 
 #define _GNU_SOURCE
+#include <errno.h>
 #include <fcntl.h>
 #include <sched.h>
 #include <string.h>
@@ -14,6 +15,7 @@
 #include <unistd.h>
 
 #include "tst_cpu.h"
+#include "tst_safe_file_ops.h"
 
 #include "util.h"
 
@@ -52,89 +54,118 @@ void burn(unsigned int usec, int sleep)
 	}
 }
 
+#define CAP_STATE_FILE_SIZE 1024
+static int read_capacity_sched_domains(int cpu, unsigned int *cap)
+{
+	char *filebuf, *tmp1, *tmp2;
+	char cap_states_file[100];
+	int cap_states_fd;
+	int bytes, rv;
+
+	sprintf(cap_states_file,
+			"/proc/sys/kernel/sched_domain/cpu%d/domain0/group0/energy/cap_states",
+			cpu);
+	cap_states_fd = open(cap_states_file, O_RDONLY);
+	if (cap_states_fd == -1)
+		return -ENOENT;
+
+	bytes = CAP_STATE_FILE_SIZE;
+	filebuf = calloc(1, CAP_STATE_FILE_SIZE + 1);
+	if (!filebuf) {
+		printf("Failed to calloc buffer for cap_states\n");
+		return -1;
+	}
+	tmp1 = filebuf;
+	while (bytes) {
+		rv = read(cap_states_fd, tmp1, bytes);
+		if (rv == -1) {
+			printf("Could not read cap_states\n");
+			return -1;
+		}
+		if (rv == 0)
+			break;
+		tmp1 += rv;
+		bytes -= rv;
+	}
+	if (tmp1 - filebuf == CAP_STATE_FILE_SIZE) {
+		printf("CAP_STATE_FILE_SIZE exhausted, increase\n");
+		return -1;
+	}
+	tmp1 = strrchr(filebuf, '\t');
+	if (!tmp1 || tmp1 == filebuf ) {
+		printf("Malformatted cap_states_file (1).\n%s\n", filebuf);
+		return -1;
+	}
+	tmp1 = strrchr(tmp1 - 1, '\t');
+	if (!tmp1 || tmp1 == filebuf) {
+		printf("Malformatted cap_states_file (2).\n%s\n", filebuf);
+		return -1;
+	}
+	tmp1 = strrchr(tmp1 - 1, '\t');
+	if (!tmp1 || tmp1 == filebuf) {
+		printf("Malformatted cap_states_file (3).\n%s\n", filebuf);
+		return -1;
+	}
+	/* tmp1 now points to tab after the capacity we want. */
+	*tmp1 = 0;
+	tmp2 = strrchr(tmp1 - 1, '\t');
+	if (!tmp2)
+		tmp2 = filebuf;
+	else
+		tmp2++;
+	if (sscanf(tmp2,"%d", cap) != 1) {
+		printf("Failed to parse capacity from cap_states.\n");
+		return -1;
+	}
+	free(filebuf);
+	if (close(cap_states_fd)) {
+		printf("Failed to close cap_states file.\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+static int read_capacity_sysfs(int cpu, unsigned int *cap)
+{
+	char path[100];
+
+	sprintf(path, "/sys/devices/system/cpu/cpu%d/cpu_capacity", cpu);
+
+	return SAFE_FILE_LINES_SCANF(path, "%u", cap);
+}
+
+static int read_cpu_capacity(int cpu, unsigned int *cap)
+{
+	int ret = read_capacity_sched_domains(cpu, cap);
+
+	/*
+	 * Capacities are exposed in sched debug for android 4.9 and older.
+	 * Otherwise, upstream uses a sysfs interface.
+	 */
+	if (ret == -ENOENT)
+		ret = read_capacity_sysfs(cpu, cap);
+
+	if (ret)
+		perror(NULL);
+
+	return ret;
+}
+
 /*
  * get_bigs = 0, search for smallest CPUs
  * get_bigs = 1, search for CPUs other than the smallest CPUs
  */
-#define CAP_STATE_FILE_SIZE 1024
 int find_cpus_with_capacity(int get_bigs, cpu_set_t *cpuset)
 {
-	int i, bytes, rv;
-	unsigned int smallest = -1;
-	char cap_states_file[100];
+	unsigned int cap, smallest = -1;
+	int i;
 
 	CPU_ZERO(cpuset);
 
 	for (i = 0; i < tst_ncpus(); i++) {
-		int cap_states_fd;
-		unsigned int cap;
-		char *filebuf, *tmp1, *tmp2;
-
-		sprintf(cap_states_file,
-			"/proc/sys/kernel/sched_domain/cpu%d/domain0/group0/energy/cap_states",
-			i);
-		cap_states_fd = open(cap_states_file, O_RDONLY);
-		if (cap_states_fd == -1) {
-			printf("Could not open cap_states file %s: ",
-			       cap_states_file);
-			perror(NULL);
-			return -1;
-		}
-		bytes = CAP_STATE_FILE_SIZE;
-		filebuf = calloc(1, CAP_STATE_FILE_SIZE + 1);
-		if (!filebuf) {
-			printf("Failed to calloc buffer for cap_states\n");
+		if (read_cpu_capacity(i, &cap))
 			return -1;
-		}
-		tmp1 = filebuf;
-		while (bytes) {
-			rv = read(cap_states_fd, tmp1, bytes);
-			if (rv == -1) {
-				printf("Could not read cap_states\n");
-				return -1;
-			}
-			if (rv == 0)
-				break;
-			tmp1 += rv;
-			bytes -= rv;
-		}
-		if (tmp1 - filebuf == CAP_STATE_FILE_SIZE) {
-			printf("CAP_STATE_FILE_SIZE exhausted, increase\n");
-			return -1;
-		}
-		tmp1 = strrchr(filebuf, '\t');
-		if (!tmp1 || tmp1 == filebuf ) {
-			printf("Malformatted cap_states_file (1).\n%s\n",
-			       filebuf);
-			return -1;
-		}
-		tmp1 = strrchr(tmp1 - 1, '\t');
-		if (!tmp1 || tmp1 == filebuf) {
-			printf("Malformatted cap_states_file (2).\n%s\n",
-			       filebuf);
-			return -1;
-		}
-		tmp1 = strrchr(tmp1 - 1, '\t');
-		if (!tmp1 || tmp1 == filebuf) {
-			printf("Malformatted cap_states_file (3).\n%s\n",
-			       filebuf);
-			return -1;
-		}
-		/* tmp1 now points to tab after the capacity we want. */
-		*tmp1 = 0;
-		tmp2 = strrchr(tmp1 - 1, '\t');
-		if (!tmp2)
-			tmp2 = filebuf;
-		else
-			tmp2++;
-		if (sscanf(tmp2,"%d", &cap) != 1) {
-		       printf("Failed to parse capacity from cap_states.\n");
-		       return -1;
-		}
-		if (close(cap_states_fd)) {
-			printf("Failed to close cap_states file.\n");
-			return -1;
-		}
 
 		if (cap < smallest) {
 			smallest = cap;
